tokio/src/runtime/task/trace/mod.rs

```diff
@@ -317,20 +319,28 @@ cfg_rt_multi_thread! {
         drop(synced);
 
         // notify each task
-        let mut traces = vec![];
+        let mut tasks = vec![];
         owned.for_each(|task| {
             // set the notified bit
             task.as_raw().state().transition_to_notified_for_tracing();
+            // store the raw tasks into a vec
+            tasks.push(task.as_raw());
+            // do NOT poll `task` here, since we hold a lock on `owned` and the task
+            // may complete and need to remove itself from `owned`.
+        });
 
-            // trace the task
-            let ((), trace) = Trace::capture(|| task.as_raw().poll());
-            traces.push(trace);
+        tasks
+            .into_iter()
+            .map(|task| {
+                // trace the task
+                let ((), trace) = Trace::capture(|| task.poll());
 
-            // reschedule the task
-            let _ = task.as_raw().state().transition_to_notified_by_ref();
-            task.as_raw().schedule();
-        });
+                // reschedule the task
+                let _ = task.state().transition_to_notified_by_ref();
+                task.schedule();
```

Comment:
I'm a few months removed from when I last hacked on this, so my memory of Tokio's internals is a bit hazy. 

These two lines are now the only difference between how owned tasks are handled on the current thread scheduler, and how they're handled on the multi thread scheduler. The current thread scheduler doesn't have these two lines, and adding them causes Tokio to panic.

It *feels* wrong for them to be missing from the current-thread scheduler: We *need* to re-schedule tasks after tracing them, right? 

Any insights here?

Reply:
There are some concurrent problems here. Did you notice that the tace all tasks implementation of `current_thread` is differenct from the implementation of `multi_thread`?

Because there are more than 1 worker threads in `multi_thread`, they could get the same tasks collection from OwnedTasks in the above `trace_multi_thread` method. Our repeated polling of the same Task in multi worker threads is not the key issue here. The key issue is that we are not allowed to poll the same task concurrently in multi worker threads.

We should find a way to avoid this concurrency issue.

Reply:
I think the tasks reschedule themselves when they are traced, so I don't think you need this.

Reply:
That said, I don't see how it would cause an error. What error do you get?

Reply:
> I think the tasks reschedule themselves when they are traced, so I don't think you need this.

Whoops, I totally missed that you added this in 7430865d655d31836ce62e65d8f03bcb9a85b12e. That's great! I'll remove these lines, then.

> That said, I don't see how it would cause an error. What error do you get?

Ah, it's a re-borrow panic of the current-thread tracing core. That makes sense.